Modules/Tokens/src/Tokens.Application/Tokens/Queries/GetToken/Handler.cs

```diff
@@ -29,9 +31,36 @@ private async Task<Token> GetToken(string tokenId, byte[]? password, Cancellatio
         var token = await _tokensRepository.Find(TokenId.Parse(tokenId), _userContext.GetAddressOrNull(), cancellationToken, true) ??
                     throw new NotFoundException(nameof(Token));
 
-        if (!token.CanBeCollectedUsingPassword(_userContext.GetAddressOrNull(), password))
-            throw new NotFoundException(nameof(Token));
+        var identity = _userContext.GetAddressOrNull();
+        if (identity == null)
+        {
+            await EnsureTokenCanBeCollected(token, null, password, cancellationToken);
+        }
+        else if (identity != token.CreatedBy)
+        {
+            if (!token.HasAllocationForIdentity(identity))
+            {
+                await EnsureTokenCanBeCollected(token, identity, password, cancellationToken);
+
+                token.AddAllocationFor(identity, _userContext.GetDeviceId());
+                await _tokensRepository.Update(token, cancellationToken);
+            }
```

Comment:
I think this stuff (except for line 46, of course) should also be placed into the `TryToAccess` method I suggested in the previous comment.